src/modules/navigator/GeofenceBreachAvoidance/geofence_breach_avoidance.cpp

```diff
@@ -0,0 +1,296 @@
+}
+
+Vector2d
+GeofenceBreachAvoidance::generateLoiterPointForFixedWing(geofence_violation_type_u violation_type, Geofence *geofence)
+{
+	if (violation_type.flags.fence_violation) {
+		const float bearing_90_left = matrix::wrap_2pi(_test_point_bearing - M_PI_F * 0.5f);
+		const float bearing_90_right = matrix::wrap_2pi(_test_point_bearing + M_PI_F * 0.5f);
+
+		double loiter_center_lat, loiter_center_lon;
+		double fence_violation_test_point_lat, fence_violation_test_point_lon;
+
+		waypoint_from_heading_and_distance(_current_pos_lat_lon(0), _current_pos_lat_lon(1), bearing_90_left,
+						   _test_point_distance, &fence_violation_test_point_lat, &fence_violation_test_point_lon);
+
+		const bool left_side_is_inside_fence = geofence->isInsidePolygonOrCircle(fence_violation_test_point_lat,
+						       fence_violation_test_point_lon, _current_alt_amsl);
+
+		waypoint_from_heading_and_distance(_current_pos_lat_lon(0), _current_pos_lat_lon(1), bearing_90_right,
+						   _test_point_distance, &fence_violation_test_point_lat, &fence_violation_test_point_lon);
+
+		const bool right_side_is_inside_fence = geofence->isInsidePolygonOrCircle(fence_violation_test_point_lat,
+							fence_violation_test_point_lon, _current_alt_amsl);
+
+		float bearing_to_loiter_point;
+
+		if (right_side_is_inside_fence && !left_side_is_inside_fence) {
+			bearing_to_loiter_point = bearing_90_right;
+
+		} else if (left_side_is_inside_fence && !right_side_is_inside_fence) {
+			bearing_to_loiter_point = bearing_90_left;
+
+		} else {
+			bearing_to_loiter_point = matrix::wrap_2pi(_test_point_bearing + M_PI_F);
+		}
+
+		waypoint_from_heading_and_distance(_current_pos_lat_lon(0), _current_pos_lat_lon(1), bearing_to_loiter_point,
+						   _test_point_distance, &loiter_center_lat, &loiter_center_lon);
+
+		return Vector2d(loiter_center_lat, loiter_center_lon);
+
+	} else if (violation_type.flags.dist_to_home_exceeded) {
+
+		return waypointFromHomeToTestPointAtDist(math::max(_max_hor_dist_home - 2 * _test_point_distance, 0.0f));
+
+	} else {
+		return _current_pos_lat_lon;
+	}
+}
+
+Vector2d
+GeofenceBreachAvoidance::generateLoiterPointForMultirotor(geofence_violation_type_u violation_type, Geofence *geofence)
+{
+
+	if (violation_type.flags.fence_violation) {
+		float current_distance = _test_point_distance * 0.5f;
+		float current_min = 0.0f;
+		float current_max = _test_point_distance;
+		Vector2d test_point;
+
+		// logarithmic search for the distance from the drone to the geofence in the given direction
+		while (abs(current_max - current_min) > 0.5f) {
+			test_point = waypointFromBearingAndDistance(_current_pos_lat_lon, _test_point_bearing, current_distance);
+
+			if (!geofence->isInsidePolygonOrCircle(test_point(0), test_point(1), _current_alt_amsl)) {
+				current_max = current_distance;
+				current_distance = current_min + (current_max - current_min) * 0.5f;
```

Comment:
This is the same in both cases -> you can move it out of the if and simplify:
```suggestion
				current_distance = (current_max + current_min) * 0.5f;
```

Reply:
When you're worried about overflow or floats that implementation can actually be more accurate, but in this case I agree it is overkill.